### PR TITLE
Generate all Seed CRDs to example

### PR DESCRIPTION
**Pull request by timebertt**
**How to categorize this PR?**

/area dev-productivity
/kind enhancement

**What this PR does / why we need it**:

With this PR, `make generate-extensions-crds` is replaced by a reusable hack script `generate-seed-crds.sh`, that can generate all of our CRDs (extensions, druid, resources) or selected groups of them using `controller-gen`.

- g/g now uses this script on `make generate` to generate all seed CRDs to `example/seed-crds` and the resource-manager CRDs to `example/resource-manager`.
- extensions can then leverage this script as well to generate the example CRDs in order to provide a simple way for developers to use the current CRDs on a development cluster (e.g. `ls example/20-crd-* | xargs -I {} kubectl apply -f {}`). See https://github.com/timebertt/gardener-extension-provider-aws/commit/a686de3e16fa46a174c846742663003772cc2604

**Which issue(s) this PR fixes**:
Part of https://github.com/gardener/gardener-extension-provider-aws/issues/430

**Special notes for your reviewer**:

Currently I'm missing CRDs for `VPA`, `HVPA` and `DNS{Provider,Owner,Entry}`.
I think, the HVPA and DNS CRDs are not strictly required as they are also not available in https://github.com/gardener/gardener-extension-provider-aws/tree/master/example.
However, there is https://github.com/gardener/gardener-extension-provider-aws/blob/master/example/20-crd-vpa.yaml but I'm not sure how we can generate it and if it's actually needed.
Any input is appreciated!

**Release note**:

```feature developer
CRDs that are installed by Gardener on a Seed cluster are now generated to `example/seed-crds`. This allows to quickly apply all Seed CRDs for development purposes.
```
```feature dependency
A new hack script `generate-seed-crds.sh` was added, that can generate all of Gardener's Seed CRDs using `controller-gen`. See [this file](https://github.com/gardener/gardener/blob/master/example/seed-crds/doc.go) for an example usage. Make sure to add `controller-gen` to the list of requirements and `example` to the list of generated paths.
```
```breaking dependency
`hack/{generate,generate-parallel.sh}` don't set `GO111MODULE=off` anymore as they used to. This was done to speed up generation with `k8s.io/code-generator`. If your repo reuses these scripts to generate code using `k8s.io/code-generator` you might want to consider setting `GO111MODULE=off` explicitly in `hack/update-codegen.sh`.
```

### CODE REVIEW

**Comment by timebertt** [comment]
Hmm, `check-generate` fails:
```
Generate github.com/gardener/gardener/example/resource-manager
controller-gen not available, please install it first by running 'make install-requirements'
example/resource-manager/doc.go:15: running "../../hack/generate-seed-crds.sh": exit status 1
Generate github.com/gardener/gardener/example/seed-crds
controller-gen not available, please install it first by running 'make install-requirements'
example/seed-crds/doc.go:15: running "../../hack/generate-seed-crds.sh": exit status 1
```
However, `controller-gen` should actually be installed. I'm confused 🤔

File: hack/generate-seed-crds.sh [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+  fi
+  local package_path="$(go list -f '{{ .Dir }}' "$package")"
+
+  # clean all generated files for this group to account for changed prefix or removed resources
+  if ls "$output_dir"/*${group}_*.yaml >/dev/null 2>&1; then
+    rm "$output_dir"/*${group}_*.yaml
+  fi
+
+  echo "Generating CRDs for $group group"
+  controller-gen crd paths="$package_path" output:crd:dir="$output_dir" output:stdout
```
Comment (timuthy) [inline hack/generate-seed-crds.sh:72]: What do you think about using a `controller-gen` bin from a dedicated `bin` directory as realized [here](https://github.com/gardener/gardener-resource-manager/pull/134/files#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R73). In the past, we had issues that different projects use different versions of `contorller-gen` and also long as we don't execute these commands in a dedicated container, we can at least place them in a git-ignored directory.

Comment (timebertt) [reply hack/generate-seed-crds.sh:72]: That's a great idea. Let's do it similarly to the yq binary, that we already install in such a way here.
I thought about doing this actually, but I didn't want to change the way controller-gen is installed in the same PR.
But now that you propose it, let's do it directly :)

Comment (timebertt) [reply hack/generate-seed-crds.sh:72]: Done in a new commit.
In a follow-up PR, we could get rid of the `$YQ` variable in `start-gardenlet` (and rather use the changed `PATH`): https://github.com/gardener/gardener/blob/27c2a5e2c4777398961867b09781cb0b8e6a9140/hack/local-development/start-gardenlet#L34 and install all other go dependency binaries in the same way as well: https://github.com/gardener/gardener/blob/27c2a5e2c4777398961867b09781cb0b8e6a9140/Makefile#L180-L184

**Comment by timebertt** [comment]
Weird, `check-generate` still [fails](https://concourse.ci.gardener.cloud/teams/gardener/pipelines/gardener-master/jobs/master-pull-request-job/builds/334#L61664a48:411:416). This time, with a different error though:
```
Generate github.com/gardener/gardener/example/seed-crds
Generating CRDs for extensions.gardener.cloud group
Generating CRDs for resources.gardener.cloud group
cannot find package "github.com/gardener/etcd-druid/api/v1alpha1" in any of:
	/usr/local/go/src/github.com/gardener/etcd-druid/api/v1alpha1 (from $GOROOT)
	/go/src/github.com/gardener/etcd-druid/api/v1alpha1 (from $GOPATH)
```

However, when I hijack the build container, I can run the following successfully:
```
/tmp/build/86a4261d# cd /go/src/github.com/gardener/gardener/
/go/src/github.com/gardener/gardener# ls -l hack/tools/bin
total 20024
-rwxr-xr-x 1 root root 20501205 Oct 15 13:12 controller-gen
/go/src/github.com/gardener/gardener# hack/tools/bin/controller-gen --version
Version: v0.7.0
/go/src/github.com/gardener/gardener# export PATH=$PWD/hack/tools/bin:$PATH
/go/src/github.com/gardener/gardener# go generate ./example/...
Generating CRDs for resources.gardener.cloud group
Generating CRDs for extensions.gardener.cloud group
Generating CRDs for resources.gardener.cloud group
Generating CRDs for druid.gardener.cloud group
/go/src/github.com/gardener/gardener# go list -f '{{ .Dir }}' github.com/gardener/etcd-druid/api/v1alpha1
/go/src/github.com/gardener/gardener/vendor/github.com/gardener/etcd-druid/api/v1alpha1
```

This is exactly how it should look like. I'm confused again ... 🤔 

EDIT:
Actually, `make check-generate` also fails with the same error on my machine, although `make generate` works perfectly fine.

**Comment by timuthy** [comment]
> Weird, check-generate still fails. This time, with a different error though:

For the Druid I recall that we first needed to `cd` into the submodule dir before we call `controller-gen` https://github.com/gardener/etcd-druid/blob/5c908089a87215c73710ee15f1a12e1397153486/Makefile#L68

**Comment by timebertt** [comment]
I think, I found the problem:

In `hack/generate.sh` and `hack/generate-parallel.sh` we set `GO111MODULE=off`: https://github.com/gardener/gardener/blob/10fe751d6960a9e764926a7002f6545f9c7cead4/hack/generate-parallel.sh#L37-L39

However, we don't even call `k8s.io/code-generator` in there – we only call it in `hack/update-{codegen,protobuf}.sh` which is invoked from `make generate`.

Because of `GO111MODULE=off`, the `go list` call in `hack/generate-seed-crds.sh` doesn't look at the vendored version of `etcd-druid`, but only looks for the package from GOROOT and GOPATH:

```
cannot find package "github.com/gardener/etcd-druid/api/v1alpha1" in any of:
	/usr/local/go/src/github.com/gardener/etcd-druid/api/v1alpha1 (from $GOROOT)
	/go/src/github.com/gardener/etcd-druid/api/v1alpha1 (from $GOPATH)
```

It wasn't failing on my machine before, because it just used my local clone of `etcd-druid` inside GOPATH.
After deleting the clone, `make generate` and `make check-generate` started failing with the same error.

### Long story short
I think we have to remove `GO111MODULE=off` in `hack/{generate,generate-parallell}.sh` (it's not used in g/g)
In https://github.com/gardener/gardener/pull/3489, @stoyanr probably added it because the scripts are used in our extensions to call `update-codegen.sh` (see https://github.com/gardener/gardener/pull/3489#discussion_r571094164).
I would vote for a) setting `GO111MODULE=off` in the scripts where it's needed (`hack/update-{codegen,protobuf}.sh`) or b) explicitly calling `hack/update-codegen.sh` with `GO111MODULE=off` from the `Makefile` in extensions.

**Comment by timebertt** [comment]
For now, I have simply dropped `GO111MODULE=off` in `hack/{generate,generate-parallell}.sh` and generated the etcd druid with the correct API version. I also added a breaking release note for dependencies reusing the hack scripts.

**Comment by timebertt** [comment]
Rebased, now that https://github.com/gardener/gardener/pull/4822 is merged.

**Comment by timebertt** [comment]
It would be good to include this in v1.34.0, given that https://github.com/gardener/gardener/pull/4851 is already merged (which will require extensions to update their CRDs, https://github.com/gardener/gardener-extension-provider-aws/issues/430). With this hack script, extension can easily regenerate new versions of their CRDs without much effort.
/milestone v1.34
cc @plkokanov

**Comment by timuthy** [comment]
Hm, `make generate` does not work for me because of some problems with the `mockgen`

```
pkg/utils/secrets/mock/doc.go:15: running "mockgen": exit status 1
Generate github.com/gardener/gardener/pkg/apis/seedmanagement/v1alpha1
I1018 13:18:37.623181    5072 main.go:127] parsing go packages in directory .
I1018 13:20:51.474834    5072 main.go:229] using package=github.com/gardener/gardener/pkg/apis/seedmanagement/v1alpha1
I1018 13:20:51.507099    5072 main.go:165] written to ../../../../hack/api-reference/seedmanagement.md
make: *** [generate] Error 40
```

Do you have the same issue?

**Comment by timebertt** [comment]
This seems to be related to https://github.com/gardener/gardener/issues/4570.
I've seen it with my PR, but also on master before.

**Review by timuthy — APPROVED** [review]
/lgtm

**Comment by plkokanov** [comment]
/lgtm

both `make generate` and `make check-generate` worked for me without a problem, further points to #4570 being the problem you guys encountered.

**Review by plkokanov — APPROVED** [review]
